### PR TITLE
#4727 amqp connection recovery

### PR DESCRIPTION
**Pull request by zuker**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -16,7 +16,7 @@
   },
   "dependencies": {
     "@elastic.io/object-storage-client": "0.0.2-dev",
-    "amqplib": "0.5.1",
+    "amqplib": "^0.6.0",
```
Comment (anton-kotenko) [inline package.json:19]: Plz no ^

File: package.json [hunk 2]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "elasticio-sailor-nodejs",
   "description": "The official elastic.io library for bootstrapping and executing for Node.js connectors",
-  "version": "2.6.18",
+  "version": "2.6.19-dev1",
```
Comment (anton-kotenko) [inline package.json:4]: Don't forget update on publish

File: run.js [hunk 3]
```diff
@@ -81,7 +85,9 @@ async function run(settings) {
 }
 
 exports._disconnectOnly = _disconnectOnly;
+exports._closeConsumerChannel = _closeConsumerChannel;
```
Comment (anton-kotenko) [inline run.js:88]: I did not found where it is used

File: package.json (hunk 1 above)
Comment (zuker) [reply package.json]: воно само! fixed

File: lib/sailor.js [hunk 4]
```diff
@@ -210,23 +210,21 @@ class Sailor {
...
-            yield this.amqpConnection.listenQueueCancel();
-            if (this.messagesCount === 0) {
-                // there is no unfinished processMessage invocation, let's just resolve scheduleShutdown now
-                log.debug('scheduleShutdown – about to shutdown immediately');
-                return;
-            }
-            // at least one processMessage invocation is not finished yet
-            // let's return a Promise, which will be resolved by processMessageAndMaybeShutdownCallback
-            log.debug('scheduleShutdown – shutdown is scheduled');
+        if (this.shutdownCallback) {
```
Comment (anton-kotenko) [inline lib/sailor.js:213]: no new logic => just kill co
?

File: lib/sailor.js [hunk 5]
```diff
@@ -456,8 +454,11 @@ class Sailor {
                     });
                 }
 
+                log.info('Going to send outgoing message');
```
Comment (anton-kotenko) [inline lib/sailor.js:457]: maybe log.trace?

File: lib/sailor.js (hunk 4 above)
Comment (zuker) [reply lib/sailor.js:213]: this.amqpConnection.listenQueueCancel() => this.amqpConnection.stopConsume();

File: lib/amqp.js [hunk 6]
```diff
@@ -48,45 +51,163 @@ class Amqp {
...
-            log.error(e,
-                'Error occurred while parsing message #%j payload',
-                amqpMessage.fields.deliveryTag
-            );
-            return this.reject(amqpMessage);
+        const amqp = await this._ensureReadConnection();
+        this.consumerChannel = await amqp.createChannel();
+        this.consumerChannel.on('error', err => log.error({ err }, 'Consumer channel error'));
+        this.consumerChannel.once('close', () => {
+            this.consumerChannel.closed = true;
```
[the remark this answers is not in the document]
Comment (zuker) [reply lib/amqp.js:100]: to ignore read async exceptions which happen when channel closed

File: lib/amqp.js [hunk 7]
```diff
@@ -48,45 +51,163 @@ class Amqp {
...
+                    if (e.isOperational && e.fields && e.fields.replyCode === 404) { // deleted queue
+                        throw e;
+                    }
+                    delete this.consumerChannel;
+                } else {
+                    throw e;
+                }
+            }
+            yield reads.filter(msg => msg);
+            await new Promise(resolve => setImmediate(resolve));
```
[the remark this answers is not in the document]
Comment (zuker) [reply lib/amqp.js:152]: tests shown that code is blocked here without setImmediate

**Review by anton-kotenko — APPROVED** [review]

**Pull request merged**